control-plane/pkg/kafka/clientpool/clientpool.go

```diff
@@ -0,0 +1,225 @@
+type GetKafkaClusterAdminFunc func(ctx context.Context, bootstrapServers []string, secret *corev1.Secret) (sarama.ClusterAdmin, ReturnClientFunc, error)
+
+func NilReturnClientFunc() {}
+
+func (cp *clientPool) get(ctx context.Context, bootstrapServers []string, secret *corev1.Secret) (sarama.Client, ReturnClientFunc, error) {
+	// (bootstrapServers, secret) uniquely identifies a sarama client config with the options we allow users to configure
+	key := makeClusterAdminKey(bootstrapServers, secret)
+
+	logger := logging.FromContext(ctx)
+
+	logger.Debug("about to get connection from clientpool", zap.Any("key", key))
+
+	// if a corresponding connection already exists, lets use it
+	if val, returnClient, ok, err := cp.Get(ctx, key); ok && err == nil {
+		logger.Debug("successfully got a client from the clientpool")
+		// check that the value is still good, as there are still some write errors
+		ca, err := cp.newClusterAdminFromClient(val)
+		if err != nil {
+			returnClient()
+			return nil, NilReturnClientFunc, err
+		}
+		if _, err := ca.ListTopics(); err != nil {
+			logger.Debug("failed to list topics with connection from clientpool, refreshing brokers and metadata")
+			err := val.RefreshBrokers(bootstrapServers)
+			if err != nil {
+				returnClient()
+				logger.Debug("failed to refresh brokers")
+				return nil, NilReturnClientFunc, err
+			}
+			err = val.RefreshMetadata()
+			if err != nil {
+				returnClient()
+				logger.Debug("failed to refresh metadata")
+				return nil, NilReturnClientFunc, err
+			}
+		}
+
+		return val, returnClient, nil
+	} else {
+		returnClient()
+		if err != nil {
+			logger.Debug("an error occurred while getting the value from the cache", zap.Error(err))
+			// the context timed out, any future actions will also fail
+			return nil, NilReturnClientFunc, fmt.Errorf("error getting an existing client: %v", err)
+		}
+	}
+
+	logger.Debug("failed to get an existing client, going to create one")
+
+	// create a new client in the client pool, and acquire capacity to start using it right away
+	saramaClient, returnClient, _, err := cp.AddAndAcquire(ctx, key, func() (sarama.Client, error) {
+		saramaClient, err := cp.makeSaramaClient(bootstrapServers, secret)
+		if err != nil {
+			return nil, err
+		}
+		return saramaClient, nil
+	}, cp.capacityPerClient)
+	if err != nil {
+		logger.Debug("failed to make a new client in the pool", zap.Error(err))
+		returnClient()
+		return nil, NilReturnClientFunc, fmt.Errorf("error creating a new client: %v", err)
+	}
+
+	return saramaClient, returnClient, nil
+}
+
+func makeClusterAdminKey(bootstrapServers []string, secret *corev1.Secret) clientKey {
+	sort.SliceStable(bootstrapServers, func(i, j int) bool {
+		return bootstrapServers[i] < bootstrapServers[j]
+	})
+
+	key := clientKey{
+		bootstrapServers: strings.Join(bootstrapServers, ","),
+	}
+
+	if secret != nil {
+		key.secretName = secret.GetName()
+		key.secretNamespace = secret.GetNamespace()
+	}
```

Comment:
I think we need to also include the "values" or the object generation / version as if I change the secret I'd like to have a new client, very relevant when creating a secret with bad data

Reply:
But then the eviction might be more complicated

Reply:
Currently, there is a event handler added to the secret informer which triggers an update of the value for any matching secret keys - do you think that is good enough?